File: 8puzzlehill.py
import copy

choice = int(input('Press 1 for Default board\nPress 2 for board with blank at 9th Cell\nPress 3 for Custom Board\n'))
if choice == 1:
    goal = {1:[0,0],2:[0,1],3:[0,2],4:[1,2],5:[2,2],6:[2,1],7:[2,0],8:[1,0],0:[1,1]}
elif choice == 2:
    goal = {1:[0,0],2:[0,1],3:[0,2],4:[1,0],5:[1,1],6:[1,2],7:[2,0],8:[2,1],0:[2,2]}
else:
    print("Enter Values for Goal state")
    goal = {}
    for i in range(3):
        goal_temp = input().split(' ')
        goal[int(goal_temp[0])] = [i,0]
        goal[int(goal_temp[1])] = [i,1]
        goal[int(goal_temp[2])] = [i,2]

print("Enter Values in 3*3 Puzzle")
initial = []
for i in range(3):
    temp = input().split(' ')
    temp = [int(i) for i in temp]
    initial.append(temp)

def calculate_manhattan(initial,goal):
    md = 0
    empty = []
    for i in range(3):
        for j in range(3):
            num = initial[i][j]
            if num == 0:
                empty.append([i,j])
            else:
                org_pos = goal.get(num)
                if org_pos[0] == i and org_pos[1] == j:
                    continue
                else:
                    md = md + abs(org_pos[0]-i) + abs(org_pos[1]-j)
            # print(num,abs(org_pos[0]-i) + abs(org_pos[1]-j))
    return md,empty[0]

def display(initial):
    for i in range(3):
        print(str(initial[i][0])+'\t'+str(initial[i][1])+"\t"+str(initial[i][2])+'\n')

def simulate(initial,goal,empty,moves,dist):
    temp_initial = []
    temp_dist = dist
    temp_empty = []
    main_move = ""
    # print(initial)
    # print(dist)
    # print('Start')
    for move in moves:
        # print(move)
        temp = copy.deepcopy(initial)
        if move == 'UP':
            zero = temp[empty[0]][empty[1]]
            temp[empty[0]][empty[1]] = temp[empty[0]+1][empty[1]]
            temp[empty[0]+1][empty[1]] = zero
        if move == 'DOWN':
            zero = temp[empty[0]][empty[1]]
            temp[empty[0]][empty[1]] = temp[empty[0]-1][empty[1]]
            temp[empty[0]-1][empty[1]] = zero
        if move == "RIGHT":
            zero = temp[empty[0]][empty[1]]
            temp[empty[0]][empty[1]] = temp[empty[0]][empty[1]+1]
            temp[empty[0]][empty[1]+1] = zero
        if move == "LEFT":
            zero = temp[empty[0]][empty[1]]
            temp[empty[0]][empty[1]] = temp[empty[0]][empty[1]-1]
            temp[empty[0]][empty[1]-1] = zero
        td,te = calculate_manhattan(temp,goal)
        # print(temp,td)
        if temp_dist > td:
            temp_dist = td
            temp_initial = copy.deepcopy(temp)
            temp_empty = te
            main_move = move
    if temp_dist == dist:
        return -1,-1,-1
    if main_move == 'DOWN':
        print("Move Number 0 UP")
    elif main_move == 'UP':
        print("Move Number 0 DOWN")
    elif main_move == "LEFT":
       print("Move Number 0 LEFT")
    elif main_move == "RIGHT":
        print("Move Number 0 RIGHT")
    display(temp_initial)
    print('-----------------------------------------------------------')
    return temp_initial,temp_dist,temp_empty

def solve(initial,goal):
    dist = 0
    dist, empty = calculate_manhattan(initial,goal)
    print('-----------------------------------------------------------')
    print('\n\n Initial State')
    display(initial)
    while(1):
        i = empty[0]
        j = empty[1]
        if (i==1) and (j == 1):
            initial,dist,empty = simulate(initial,goal,empty,['UP','DOWN','LEFT','RIGHT'],dist)
        elif (i==0 and j==0) or (i==0 and j==2) or (i==2 and j == 2) or (i==2 and j==0):
            if i==0:
                if j == 0:
                    initial,dist,empty = simulate(initial,goal,empty,['UP','RIGHT'],dist)
                else:
                    initial,dist,empty = simulate(initial,goal,empty,['UP','LEFT'],dist)
            else:
                if j == 2:
                    initial,dist,empty = simulate(initial,goal,empty,['DOWN','LEFT'],dist)
                else:
                    initial,dist,empty = simulate(initial,goal,empty,['DOWN','RIGHT'],dist)
        else:
            if i == 0:
                initial,dist,empty = simulate(initial,goal,empty,['UP','LEFT','RIGHT'],dist)
            elif i ==1:
                if j == 0:
                    initial,dist,empty = simulate(initial,goal,empty,['UP','DOWN','RIGHT'],dist)
                else:
                    initial,dist,empty = simulate(initial,goal,empty,['UP','DOWN','LEFT'],dist)
            else:
                initial,dist,empty = simulate(initial,goal,empty,['DOWN','LEFT','RIGHT'],dist)
        if initial == -1:
            print("No solution Possible")
            break
        if dist == 0:
            break
solve(initial,goal)

File: 8puzzlehill.cpp
#include <bits/stdc++.h>

using namespace std;

const char UP ='1';
const char DOWN= '2';
const char LEFT= '3';
const char RIGHT= '4';


void printArray(int** array)
{
    for(int a=0;a<3;a++)
    {
        for(int b=0;b<3;b++)
        {
      		cout<< setw(8) << array[a][b];    
        }
        cout<<endl;
    }
    cout<<endl;
}

// Calculate Manhattan distance
int manhattan_distance(int** start_state,int** goal_state)
{
    int manhattan_distance=0;
    for(int i=0;i<3;i++)
        {
            for(int j=0;j<3;j++)
            {
                if(start_state[i][j] > 0)
                {
                    for(int k=0;k<3;k++)
                    {
                        for(int l=0;l<3;l++)
                        {
                            if (goal_state[k][l] == start_state[i][j])
                            {
                                manhattan_distance=manhattan_distance + (abs(i-k)+abs(j-l));
                            
                            }
                        }
                    }
                }
            }
        }
return manhattan_distance;

}

void makeMove(int** temp,int move)
{
    int flag=0;
    for(int i=0;i<3;i++)
    {
        for(int j=0;j<3;j++)
        {
            if(temp[i][j] == 0)
            {
         
            
                if(move==1)
                {
                    temp[i][j] = temp[i-1][j];
                	temp[i-1][j] = 0;
                    flag=1;
                    break;
                }
                else if(move==2)
                {     
                    temp[i][j] = temp[i+1][j];
                	temp[i+1][j] = 0;
                    flag=1;
                    break;
                }
                else if(move==3)

                {     
                    temp[i][j] = temp[i][j-1];
                	temp[i][j-1] = 0;
                    flag=1;
                    break;
                }
                else
                {
                	temp[i][j] = temp[i][j+1];
                	temp[i][j+1] = 0;
                    flag=1;
                    break;
                }  
            }
        }
        if(flag==1){break;}
    }
}
   
//----------------------------------------------------------------------------------------------------------- 
int tile_Ordering(int** current_state,int** goal_state,int move) 
{
    int** temp = new int*[3]; // allocate an array of 3 int pointers - these represents rows
    for(int i=0;i<3;i++)
    {
        temp[i]=new int[3]; // these represents columns
        for(int j=0;j<3;j++)
        {
        	temp[i][j] = current_state[i][j];
        }
    }
    makeMove(temp,move);
    printArray(temp);
    int manhattan=manhattan_distance(temp,goal_state);
    cout<<"Current Manhattan number :"<<manhattan<<endl<<endl<<endl;
    for(int i=0;i<3;i++)
    {
        delete temp[i];
    }
    delete temp;
    return manhattan; 
}

//----------------------------------------------------------------------------------------------------------
void HillClimbing( int** start_state, int** goal_state,int former_move)
{
	int arr[4] = {100,100,100,100};
    cout<<"--------------------------------------------------------------------------------"<<endl;
    for(int i=0;i<3;i++)
    {
        for(int j=0;j<3;j++)
        {
            if (start_state[i][j] == 0)
            {
                if(i>0 && former_move!=2)  
                {
                    cout<<"Checking child (moving 0 up) "<<endl;
                    arr[0]=tile_Ordering(start_state,goal_state,1); 
                }
                if(i<2 && former_move!=1) 
                {
                    cout<<"Checking child (moving 0 down) "<<endl;
                    arr[1]=tile_Ordering(start_state,goal_state,2);
                }
                if(j>0 && former_move!=4)
                {
                    cout<<"Checking child (moving 0 left) "<<endl;
                    arr[2]=tile_Ordering(start_state,goal_state,3);
                }
                if(j<2 && former_move!=3)
                {
                    cout<<"Checking child (moving 0 right) "<<endl;
                    arr[3]=tile_Ordering(start_state,goal_state,4);
                }
            }
        }
        cout<<endl;
    }
    int localOptimum=99;
    int index=0;
    for(int i=0;i<4;i++)
    {
        if(arr[i]<localOptimum)
        {
            localOptimum=arr[i];
            index=i+1;
        }
    }
    makeMove(start_state,index);
    cout<<"Next state = minimum Manhattan number :"<<endl;
    printArray(start_state);
    if(localOptimum==0)
    {  
        cout<<"goal state reached"<<endl;

        return;
    }
    else
    HillClimbing(start_state,goal_state,index);
}

int main()
{
    int** initial = new int*[3]; // allocate an array of 3 int pointers - these represents rows
    for(int i=0;i<3;i++)
    {
        initial[i]=new int[3]; // these represents columns
    }

    int** final = new int*[3]; // allocate an array of 3 int pointers - these represents rows
    for(int i=0;i<3;i++)
    {
        final[i]=new int[3]; // these represents columns
    }

    int player_Input;
    cout << "Enter initial board configuration - 0 denotes empty position" << endl;
    for(int i=0;i<3;i++)
    {
        for(int j=0;j<3;j++)
        {
            cout<<"Enter input A["<<i<<"]["<<j<<"]"<<endl;
            cin>>player_Input;
            initial[i][j]=player_Input;
        }
    }
    
    cout<<"--------------------------------------------------------------------------------"<<endl;
    cout << "Enter final board configuration - 0 denotes empty position" << endl;
    for(int i=0;i<3;i++)
        for(int j=0;j<3;j++)
        {
            cout<<"Enter input A["<<i<<"]["<<j<<"]"<<endl;
            cin>>player_Input;
            final[i][j]=player_Input;
        }

    cout<<"\n---------------------------Your initial matrix is-------------------------------\n"<<endl;	
  

    for(int i=0;i<3;i++)
    {
        for(int j=0;j<3;j++)
        {
        	cout << setw(8) << initial[i][j];	     
        }
        cout<<endl;
    }
    
    cout<<"\n---------------------------Your final matrix is--------------------------------\n"<<endl;
  
    
    for(int i=0;i<3;i++)
    {
        for(int j=0;j<3;j++)
        {
        	cout << setw(8) << final[i][j];  
        }
        cout<<endl;
    } 

	cout<<"\n--------------------------------------------------------------------------------"<<endl;
	cout<<"\n--------------------------------------------------------------------------------"<<endl;
    cout<<"\nCalling Hill Climbing function\n"<<endl;
   HillClimbing(initial,final,0);
    for(int i=0;i<3;i++)
    {
        delete initial[i];
        delete final[i];
    }
    delete initial;
    delete final;
    return 0;
}
